lib/interactive_viewer.ml

```diff
@@ -115,44 +115,163 @@ let toggle_view_mode () =
+      let empty_mine = fill_empty [] (their_len - mine_len) in
+      ( List.rev (empty_mine @ temp_mine),
+        List.rev temp_their,
+        mine_hunk_length,
+        their_hunk_length )
+    else
+      ( List.rev temp_mine,
+        List.rev temp_their,
+        mine_hunk_length,
+        their_hunk_length )
+  in
+
   match hunks with
-  | [] -> (List.rev mine_acc, List.rev their_acc)
-  | (`Common _ as common) :: t ->
-      split_and_align_hunk t (common :: mine_acc) (common :: their_acc)
-  | `Mine s :: `Their s' :: t ->
-      split_and_align_hunk t (`Mine s :: mine_acc) (`Their s' :: their_acc)
-  | `Mine s :: t ->
-      split_and_align_hunk t (`Mine s :: mine_acc) (`Common "" :: their_acc)
-  | `Their s :: t ->
-      split_and_align_hunk t (`Common "" :: mine_acc) (`Their s :: their_acc)
-
-let lines_to_ui lines attr_change =
-  List.map
-    (fun line ->
-      let content, attr =
-        match line with
-        | `Common s -> (s, Notty.A.empty)
-        | `Mine s -> (s, attr_change)
-        | `Their s -> (s, attr_change)
+  | [] ->
+      ( List.rev mine_acc,
+        List.rev their_acc,
+        mine_hunk_length,
+        their_hunk_length )
+  | _ -> (
+      let ( temp_mine,
+            temp_their,
+            new_mine_hunk_length,
+            new_their_hunk_length,
+            remaining_hunks ) =
+        buffer_changes [] [] mine_hunk_length their_hunk_length hunks
+      in
+      let ( balanced_mine,
+            balanced_their,
+            final_mine_hunk_length,
+            final_their_hunk_length ) =
+        append_balanced temp_mine temp_their new_mine_hunk_length
+          new_their_hunk_length
       in
-      W.string ~attr content)
-    lines
+      let updated_mine_acc = List.rev_append balanced_mine mine_acc in
+      let updated_their_acc = List.rev_append balanced_their their_acc in
+      match remaining_hunks with
+      | `Common line :: t ->
+          let common_mine_acc = Common line :: updated_mine_acc in
+          let common_their_acc = Common line :: updated_their_acc in
+          split_and_align_hunk t common_mine_acc common_their_acc
+            final_mine_hunk_length final_their_hunk_length
+      | _ ->
+          split_and_align_hunk remaining_hunks updated_mine_acc
+            updated_their_acc final_mine_hunk_length final_their_hunk_length)
+
+let lines_with_numbers lines attr_change line_num prefix =
+  List.fold_left
+    (fun acc line ->
+      match line with
+      | Common s ->
+          line_num := !line_num + 1;
+          let content = Printf.sprintf "%3d   %s" !line_num s in
+          (content, Notty.A.empty) :: acc
+      | Change (Mine, s) ->
+          incr line_num;
+          let content = Printf.sprintf "%3d %s %s" !line_num prefix s in
+          (content, attr_change) :: acc
+      | Change (Their, s) ->
+          incr line_num;
+          let content = Printf.sprintf "%3d %s %s" !line_num prefix s in
+          (content, attr_change) :: acc
```

Comment:
What I meant here is that the `change_origin` type is constructed several times but never matched (in a meaningful way). It could be removed entirely.